src/functional/ffront/decorator.py

```diff
@@ -0,0 +1,227 @@
+    foast_node: foast.FieldOperator
+    closure_refs: ClosureRefs
+    externals: dict[str, Any]
+    backend: str  # note: backend is only used if directly called
+    definition: Optional[types.FunctionType] = None
+
+    @classmethod
+    def from_function(
+        cls, definition: types.FunctionType, externals: Optional[dict] = None, backend: str = None
+    ):
+        closure_refs = ClosureRefs.from_function(definition)
+        foast_node = FieldOperatorParser.apply_to_function(definition)
+        return FieldOperator(
+            foast_node=foast_node,
+            closure_refs=closure_refs,
+            externals={} if externals is None else externals,
+            backend=backend,
+            definition=definition,
+        )
+
+    def __gt_type__(self) -> ct.FunctionType:
+        type_ = symbol_makers.make_symbol_type_from_value(self.definition)
+        assert isinstance(type_, ct.FunctionType)
+        return type_
+
+    def __gt_itir__(self):
+        return FieldOperatorLowering.apply(self.foast_node)
+
+    def as_program(self) -> Program:
+        if any(param.id == "out" for param in self.foast_node.params):
+            raise Exception(
+                "Direct call to Field operator whose signature contains an argument `out` is not permitted."
+            )
+
+        name = self.foast_node.id
+        loc = self.foast_node.location
+
+        type_ = self.__gt_type__()
+        stencil_sym = past.Symbol(id=name, type=type_, namespace=ct.Namespace.CLOSURE, location=loc)
+
+        params_decl = [
+            past.Symbol(id=gensym(), type=arg_type, namespace=ct.Namespace.LOCAL, location=loc)
+            for arg_type in type_.args
+        ]
+        params_ref = [past.Name(id=pdecl.id, location=loc) for pdecl in params_decl]
+        out_sym = past.Symbol(
+            id="out", type=type_.returns, namespace=ct.Namespace.LOCAL, location=loc
+        )
+        out_ref = past.Name(id="out", location=loc)
+
+        untyped_past_node = past.Program(
+            id=f"__field_operator_{name}",
+            params=params_decl + [out_sym],
+            body=[
+                past.Call(
+                    func=past.Name(id=name, location=loc),
+                    args=params_ref,
+                    kwargs={"out": out_ref},
+                    location=loc,
+                )
+            ],
+            closure=[stencil_sym],
+            location=loc,
+        )
+        past_node = ProgramTypeDeduction.apply(untyped_past_node)
+
+        # inject stencil as a closure var into program
+        #  since ClosureRefs is immutable we have to resort to this rather ugly way of doing a copy
+        closure_refs = ClosureRefs(
+            **{
+                **{
+                    field.name: getattr(self.closure_refs, field.name)
+                    for field in dataclasses.fields(self.closure_refs)
+                },
+                "globals": {**self.closure_refs.globals, name: self},
+            }
+        )
```

Comment:
Did you try with `dataclasses.replace()` https://devdocs.io/python~3.10/library/dataclasses#dataclasses.replace ?

Reply:
Also, is the reason for ClosureRefs being immutable really stronger than the inconvenience it creates here?  Is it really difficult to avoid changing it when it shouldn't be in the normal course of things?

Reply:
@egparedes That is exactly what I was looking for. Works great.
@DropD Good point. IMO it makes sense for ClosureRefs to be immutable and with `replace` the inconvenience is managable.